### ISSUE TITLE
ipld/unixfs: Dagreader ignores internal data node when links are present

### ISSUE THREAD

**Issue by ellttBen** [ipfs/boxo#390]
Hello!

I'm currently studying a use case for go-ipfs in which a file node contains links to other versions of itself. 
This all works well until the root file becomes small enough to fit inside the root's data node. Then, accessing the file through `ipfs cat` or the gateway skips the data node and goes on to read from the child nodes.
After investigating the bug, I found that it originated from `go-unixfs`, and especially the dagreader.
More precisely, https://github.com/ipfs/go-unixfs/blob/68c015a6f317ed5e21a4870f7c423a4b38b90a96/io/dagreader.go#L175-L179

I already have a fix for this if there is any interest, although it relies on the empirical assumption that we should stop descending through the DAG when we read a node that has *named* links (`len(node.Links()[0].Name) != 0`).

This would also close https://github.com/ipfs/go-ipfs/issues/7190 , which I opened a month ago.

**Comment by lidel** [comment]
## 2026 outlook on this

The use case described here (named links on a File node) is not valid UnixFS.
The [UnixFS spec](https://specs.ipfs.tech/unixfs/#pbnode-links-name) explicitly requires parsers to reject file chunks with non-empty `Name` fields.

The root cause [`ipfs object patch add-link` silently creating invalid DAGs] is fixed in
- https://github.com/ipfs/kubo/pull/11248. 
The command now rejects adding named links to non-directory nodes.

The dagreader's behavior of skipping `Data` on nodes with `Links` is correct for all valid UnixFS structures: standard builders (balanced, trickle) never  produce File nodes with both inline data bytes and child links. 

While the spec does [allow inline data before links](https://specs.ipfs.tech/unixfs/#decode-pbnode-data-data) in theory, but no builder produces such nodes and fixing the reader for this edge case is not justified given the complexity (especially in `Seek` offset calculations). 

If a real-world need arises, it can be revisited, but unlikely within `dag-pb`.

## TLDR (remaining decision here)

I guess we should keep this open to discuss if we want to make UnixFS spec more strict and error on data in intermediate nodes, or if we want to pay complexity  tax and add support for it.

If anyone has opinion, need or use case in mind, please comment below.
